tools/gomod-local-update/internal/updater/module_operator.go

```diff
@@ -0,0 +1,154 @@
+	if !remoteRE.MatchString(remote) {
+		return fmt.Errorf("%w: invalid git remote format", ErrModOperation)
+	}
+
+	branchRE := regexp.MustCompile(gitBranchPattern)
+	if !branchRE.MatchString(branch) {
+		return fmt.Errorf("%w: invalid git branch format", ErrModOperation)
+	}
+	return nil
+}
+
+// GetGitInfo retrieves the latest commit SHA and timestamp from a Git repository
+func (m *moduleOperator) GetGitInfo(remote, branch string) (string, time.Time, error) {
+	// Validate remote and branch against strict regex patterns before using in exec
+	if err := validateGitInput(remote, branch); err != nil {
+		return "", time.Time{}, err
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), gitTimeout)
+	defer cancel()
+
+	// Safe to use remote/branch after validateGitInput ensures they match:
+	// - remote: ^[a-zA-Z0-9][-a-zA-Z0-9_.]*$
+	// - branch: ^[a-zA-Z0-9][-a-zA-Z0-9/_]*$
+	//nolint:gosec // Inputs are validated by regex patterns above
+	cmd := exec.CommandContext(ctx, "git", "ls-remote", remote, "refs/heads/"+branch)
+	out, err := cmd.Output()
+	if err != nil {
+		return "", time.Time{}, fmt.Errorf("%w: failed to get SHA: %w", ErrModOperation, err)
+	}
+	if len(out) == 0 {
+		return "", time.Time{}, fmt.Errorf("%w: no output from git ls-remote", ErrModOperation)
+	}
+	sha := strings.Split(string(out), "\t")[0]
+	if len(sha) == 0 {
+		return "", time.Time{}, fmt.Errorf("%w: empty SHA from git ls-remote", ErrModOperation)
+	}
+
+	cmd = exec.CommandContext(ctx, "git", "show", "-s", "--format=%cI", sha)
+	out, err = cmd.Output()
+	if err != nil {
+		return "", time.Time{}, fmt.Errorf("failed to get commit time: %w", err)
+	}
+	if len(out) == 0 {
+		return "", time.Time{}, fmt.Errorf("%w: no output from git show", ErrModOperation)
+	}
+
+	commitTime, err := time.Parse(gitTimeFormat, strings.TrimSpace(string(out)))
+	if err != nil {
+		return "", time.Time{}, fmt.Errorf("failed to parse commit time: %w", err)
+	}
+
+	return sha[:gitSHALength], commitTime, nil
+}
+
+// GetLatestVersion retrieves the latest pseudo-version for a module
+func (m *moduleOperator) GetLatestVersion(modulePath string) (module.Version, error) {
+	sha, commitTime, err := m.GetGitInfo(m.config.RepoRemote, m.config.BranchTrunk)
+	if err != nil {
+		return module.Version{}, err
+	}
+
+	majorVer := strings.TrimPrefix(getMajorVersion(modulePath), "v")
+	pseudoVersion := module.PseudoVersion("v"+majorVer, "", commitTime, sha)
+
+	return module.Version{
+		Path:    modulePath,
+		Version: pseudoVersion,
+	}, nil
+}
+
+// extractVersionParts splits a pseudo-version into its components
+func extractVersionParts(version string) (base, timestamp, sha string) {
+	parts := strings.Split(version, "-")
+	if len(parts) == 3 {
+		return parts[0], parts[1], parts[2]
+	}
+	return "", "", ""
+}
+
+// UpdateRequiredVersions identifies modules that need version updates
+func (m *moduleOperator) UpdateRequiredVersions(modFile *modfile.File, newVersion string) ([]string, error) {
```

Comment:
Where/how is this method used? I see `ModulesToUpdate` in the config, but that seems to be populated by `updater.findLocalReplaceModules`. I don't see any direct reference to this method.

Reply:
If this is indeed used and I just missed it, I think it could probably have a better name.

Reply:
I think this was a relic of a refactor. Good catch, removed in: a2b1834b52303b7d5483bc639d8e73c8aec98ba4